### PR TITLE
fix(bug): Added mutext for downloader providers. Fixes #1531

### LINKED ISSUE #1531: Multiple goroutines access the same map in provider.go
/kind bug

**What steps did you take and what happened:**
[A clear and concise description of what the bug is.]
1. Deployed inference service
2. Deployed 2 trained models with s3 protocol

**What did you expect to happen:**
The files will be loaded normally as the first call to getProvider creates the S3 provider whereas the second call will use the already made one.

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]
A simple lock should fix this.
```
fatal error: concurrent map writes

goroutine 136 [running]:
runtime.throw(0x1e28a84, 0x15)
        runtime/panic.go:1116 +0x72 fp=0xc0009a38f0 sp=0xc0009a38c0 pc=0x437772
runtime.mapassign_faststr(0x1b60680, 0xc000788600, 0x1e0fea4, 0x5, 0x1)
        runtime/map_faststr.go:211 +0x3f1 fp=0xc0009a3958 sp=0xc0009a38f0 pc=0x4147b1
github.com/kubeflow/kfserving/pkg/agent/storage.GetProvider(0xc000788600, 0x1e0fea4, 0x5, 0x5, 0x0, 0x0, 0xc00067e410)
        github.com/kubeflow/kfserving/pkg/agent/storage/utils.go:130 +0x449 fp=0xc0009a3b40 sp=0xc0009a3958 pc=0x18690c9
github.com/kubeflow/kfserving/pkg/agent.(*Downloader).download(0xc000098c08, 0xc000036350, 0x6, 0xc0000303a0, 0x1f, 0xc000902b70, 0x2)
        github.com/kubeflow/kfserving/pkg/agent/downloader.go:80 +0x85 fp=0xc0009a3bc8 sp=0xc0009a3b40 pc=0x1956b85
github.com/kubeflow/kfserving/pkg/agent.(*Downloader).DownloadModel(0xc000098c08, 0xc000036350, 0x6, 0xc00004c840, 0x0, 0x0)
        github.com/kubeflow/kfserving/pkg/agent/downloader.go:49 +0x4bf fp=0xc0009a3d20 sp=0xc0009a3bc8 pc=0x195677f
github.com/kubeflow/kfserving/pkg/agent.(*Puller).modelProcessor(0xc000098be0, 0xc000036350, 0x6, 0xc00048c420)
        github.com/kubeflow/kfserving/pkg/agent/puller.go:144 +0x275 fp=0xc0009a3fc0 sp=0xc0009a3d20 pc=0x1957dd5
runtime.goexit()
        runtime/asm_amd64.s:1374 +0x1 fp=0xc0009a3fc8 sp=0xc0009a3fc0 pc=0x46fe61
created by github.com/kubeflow/kfserving/pkg/agent.(*Puller).enqueueModelOp
        github.com/kubeflow/kfserving/pkg/agent/puller.go:99 +0x125

goroutine 1 [select, 1 minutes]:
main.main()
        github.com/kubeflow/kfserving/cmd/agent/main.go:188 +0x8bd

goroutine 18 [chan receive]:
k8s.io/klog/v2.(*loggingT).flushDaemon(0x2e14f20)
        k8s.io/klog/v2@v2.2.0/klog.go:1131 +0x8b
created by k8s.io/klog/v2.init.0
        k8s.io/klog/v2@v2.2.0/klog.go:416 +0xd8

goroutine 19 [chan receive]:
k8s.io/klog.(*loggingT).flushDaemon(0x2e14e40)
        k8s.io/klog@v1.0.0/klog.go:1010 +0x8b
created by k8s.io/klog.init.0
        k8s.io/klog@v1.0.0/klog.go:411 +0xd8

```

### PR DESCRIPTION
**Pull request by NikeNano**
**What this PR does / why we need it**:

This PR aims to fix: https://github.com/kubeflow/kfserving/issues/1531
**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #1531

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Release note**:

```release-note

```

### CODE REVIEW

File: pkg/agent/downloader.go [hunk 1]
```diff
@@ -29,10 +29,12 @@ import (
 	"path/filepath"
 	"regexp"
 	"strings"
+	"sync"
 )
 
 type Downloader struct {
 	ModelDir  string
+	mu        sync.Mutex
```
Comment (abchoo) [inline pkg/agent/downloader.go:37]: By the way, if you were to change this to `sync.Mutex` then calling the lock would simply be `d.Lock()`.

Comment (NikeNano) [reply pkg/agent/downloader.go]: Thanks, still new to go so never used this feature before.

Comment (abchoo) [reply pkg/agent/downloader.go]: We all start somewhere :)

File: pkg/agent/downloader.go [hunk 2]
```diff
@@ -20,19 +20,22 @@ import (
...
+	"github.com/kubeflow/kfserving/pkg/agent/storage"
+	"github.com/kubeflow/kfserving/pkg/apis/serving/v1alpha1"
+	"github.com/pkg/errors"
+	"go.uber.org/zap"
 )
 
 type Downloader struct {
-	ModelDir  string
+	ModelDir string
+	sync.Mutex
```
Comment (yuzisun) [inline pkg/agent/downloader.go:38]: I actually prefer unexported type(`mu sync.Mutex`) here, embedded mutex struct is useful when you have lock/unlock as public API, but here I think lock is only used internally without exposing.

Comment (NikeNano) [reply pkg/agent/downloader.go]: I see your point and have updated.

**Comment by NikeNano** [comment]
Updated it @yuzisun

**Comment by NikeNano** [comment]
`Python package / build (3.7)` seems to fail for some reason ...

**Pull request merged**